File: mcp-gateway/revocation.py
"""Token revocation / blocklist for MCP Gateway — Valkey-backed.

Supports:
- Revoking individual JWTs by jti
- Revoking all tokens for an identity (identity_id-based)
- Checking revocation status on every auth

Keys:
- mcp:rev:jti:{jti} — individual JWT revocation (TTL = token remaining lifetime)
- mcp:rev:identity:{identity_id} — identity-level revocation (timestamp cutoff)
"""
import logging
import os
import time
from typing import Optional

logger = logging.getLogger(__name__)

_client = None


def _get_client():
    """Return a Valkey client, or None if unavailable."""
    global _client
    if _client is not None:
        return _client
    try:
        from valkey import Valkey
        host = os.environ.get("VALKEY_HOST", "valkey")
        port = int(os.environ.get("VALKEY_PORT", "6379"))
        password = os.environ.get("VALKEY_PASSWORD") or None
        client = Valkey(
            host=host, port=port, db=0, password=password,
            socket_connect_timeout=1, socket_timeout=1,
            decode_responses=True,
        )
        client.ping()
        _client = client
        return client
    except Exception as e:
        logger.debug("Valkey not available for revocation: %s", e)
        _client = None
        return None


def revoke_jti(jti: str, ttl_seconds: int = 3600) -> bool:
    """Revoke a specific JWT by its jti claim. Returns True if stored."""
    client = _get_client()
    if not client or not jti:
        return False
    try:
        client.setex(f"mcp:rev:jti:{jti}", ttl_seconds, "1")
        return True
    except Exception as e:
        logger.warning("Failed to revoke jti: %s", e)
        return False


def revoke_identity(identity_id: int) -> bool:
    """Revoke all tokens for an identity. Returns True if stored."""
    client = _get_client()
    if not client:
        return False
    try:
        client.set(f"mcp:rev:identity:{identity_id}", str(time.time()))
        # Also kill all sessions for this identity
        _revoke_identity_sessions(identity_id)
        return True
    except Exception as e:
        logger.warning("Failed to revoke identity: %s", e)
        return False


def _revoke_identity_sessions(identity_id: int) -> None:
    """Delete all sessions belonging to an identity."""
    client = _get_client()
    if not client:
        return
    try:
        for key in client.scan_iter("mcp:sess:*"):
            raw = client.get(key)
            if raw:
                import json
                data = json.loads(raw)
                if data.get("identity_id") == identity_id:
                    client.delete(key)
    except Exception as e:
        logger.warning("Failed to revoke sessions for identity %d: %s", identity_id, e)


def is_jti_revoked(jti: str) -> bool:
    """Check if a JWT jti has been revoked."""
    if not jti:
        return False
    client = _get_client()
    if not client:
        return False
    try:
        return client.exists(f"mcp:rev:jti:{jti}") > 0
    except Exception:
        return False


def is_identity_revoked(identity_id: int) -> bool:
    """Check if an identity has been revoked at the identity level."""
    client = _get_client()
    if not client:
        return False
    try:
        return client.exists(f"mcp:rev:identity:{identity_id}") > 0
    except Exception:
        return False


def is_token_valid(identity_id: int, jti: Optional[str] = None) -> bool:
    """Check if a token is still valid (not revoked).

    Returns True if valid, False if revoked.
    If Valkey is unavailable, returns True (fail-open for availability).
    """
    client = _get_client()
    if not client:
        return True
    if is_identity_revoked(identity_id):
        return False
    if jti and is_jti_revoked(jti):
        return False
    return True

File: mcp-gateway-rs/crates/policy/src/revocation.rs
//! Token revocation / blocklist — Valkey-backed, `mcp:gw:` prefixed.
//!
//! Mirrors `mcp-gateway/revocation.py` semantics:
//! - `mcp:gw:rev:jti:{jti}` — individual JWT revocation (TTL = token lifetime)
//! - `mcp:gw:rev:identity:{id}` — identity-level revocation (timestamp cutoff)
//!
//! Fail-open when Valkey is unavailable (availability over strictness, matching
//! the Python gateway).

use crate::valkey::ValkeyClient;

#[derive(Clone)]
pub struct RevocationStore {
    client: Option<ValkeyClient>,
}

impl RevocationStore {
    pub fn new(client: Option<ValkeyClient>) -> Self {
        Self { client }
    }

    /// Revoke a specific JWT by its jti claim.
    pub async fn revoke_jti(&self, jti: &str, ttl_seconds: u64) -> bool {
        let Some(c) = &self.client else { return false };
        match c.setex(&format!("rev:jti:{jti}"), ttl_seconds, "1").await {
            Ok(_) => true,
            Err(e) => {
                ValkeyClient::log_failure("revoke_jti", &e);
                false
            }
        }
    }

    /// Revoke all tokens for an identity.
    pub async fn revoke_identity(&self, identity_id: i64) -> bool {
        let Some(c) = &self.client else { return false };
        let ts = chrono::Utc::now().timestamp();
        match c.set(&format!("rev:identity:{identity_id}"), &ts.to_string()).await {
            Ok(_) => true,
            Err(e) => {
                ValkeyClient::log_failure("revoke_identity", &e);
                false
            }
        }
    }

    /// Check if a JWT jti has been revoked.
    pub async fn is_jti_revoked(&self, jti: &str) -> bool {
        if jti.is_empty() {
            return false;
        }
        let Some(c) = &self.client else { return false };
        c.exists(&format!("rev:jti:{jti}")).await.unwrap_or(false)
    }

    /// Check if an identity has been revoked at the identity level.
    pub async fn is_identity_revoked(&self, identity_id: i64) -> bool {
        let Some(c) = &self.client else { return false };
        c.exists(&format!("rev:identity:{identity_id}")).await.unwrap_or(false)
    }

    /// Check if a token is still valid (not revoked). Returns `true` if valid.
    /// Fail-open (returns true) when Valkey is unavailable.
    pub async fn is_token_valid(&self, identity_id: i64, jti: Option<&str>) -> bool {
        if self.client.is_none() {
            return true;
        }
        if self.is_identity_revoked(identity_id).await {
            return false;
        }
        if let Some(j) = jti {
            if self.is_jti_revoked(j).await {
                return false;
            }
        }
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn fail_open_without_valkey() {
        let store = RevocationStore::new(None);
        // No Valkey → fail-open (token considered valid).
        assert!(store.is_token_valid(1, Some("jti-1")).await);
        assert!(!store.revoke_jti("jti-1", 3600).await);
        assert!(!store.revoke_identity(1).await);
    }
}
